vars/checkPatch.groovy

```diff
@@ -35,12 +55,15 @@ def call(Map config) {
                  description: env.STAGE_NAME,
                  context: "pre-build" + "/" + env.STAGE_NAME,
                  status: "PENDING"
-
-    script = 'GH_USER=' + config['user'] + \
-             ' GH_PASS=' + config['password'] + \
-             ' jenkins/code_review/jenkins_github_checkwarn.sh'
-    rc = sh(script: script, returnStatus: true)
-
+    withCredentials([[$class: 'UsernamePasswordMultiBinding', 
+                    credentialsId: 'aa4ae90b-b992-4fb6-b33b-236a53a26f77',
```

Comment:
I'm not sure I like embedding the `credentialsId` in here vs. passing the username/password.  It feels like it's removing some of the generic/portability of this function.

Yes, we do have the Gerrit credentials above also, but that is just temporary until we get the `code_review/` scripts out onto GitHub.

Reply:
My point is that this is a generic library function.  There is no reason that this function needs to know anything about Jenkins credentials.  It simply needs the GitHub account credentials which can be from a Jenkins credential or otherwise.

Using a Jenkins credential here makes this otherwise generic function very specific to the way we are using Jenkins.

Reply:
And my point is that if we resolve fetch the credentials in the Jenkinsfile, particularly as environment variables, they have a larger chance of accidentally leaking into log file.

So either this method needs to have a credential ID with a generic name stored in it, or it needs to be called with a credential ID as a parameter.

We should not ever be passing in usernames and passwords to library methods.

Since you do not want to put a specific credential ID in this library, that leaves passing the credential ID.

Reply:
But we already know that Jenkins masks these account values.

```
[checkpatch] Running script: GH_USER="****" GH_PASS="****" CHECKPATCH_IGNORED_FILES="code_review/checkpatch.pl::src/control/vendor/*" jenkins/code_review/jenkins_github_checkwarn.sh
[Pipeline] [checkpatch] sh
[checkpatch] [daos-stack-org_daos_PR-13-XSYVVAKVRRCIRZBUBQLEPYMVXQHYAQV52HHPQR55EURLM5NXUHMA@2] Running shell script
[checkpatch] + GH_USER=****
[checkpatch] + GH_PASS=****
[checkpatch] + CHECKPATCH_IGNORED_FILES='code_review/checkpatch.pl::src/control/vendor/*'
[checkpatch] + jenkins/code_review/jenkins_github_checkwarn.sh
```

Reply:
Decoding the credentials into global environment variables is a bad practice.

They are masked there in the Jenkins log, but the environment variables are globally accessible to all steps in the Jenkinsfle and could get dumped in logs/artifacts that Jenkins does not know how to censor.

Decoded credentials should be restricted only to location where they are used as much as possible.

If a library method needs a password or equivalent token, we should pass it the credential id for it to get the access.

Reply:
I just noticed that the 'daos-jenkins-commit-status' ID is also present in this module.

Does the  'daos-jenkins-commit-status' credential ID used to update the status also have the ability to annotate the code?  If so, then there is a simple solution just to use it.

Reply:
But there is no need for the library function to need to understand or work with Jenkins credentials rather than the more generic form that a credential will yield.  Using the credential id makes the library function less flexible to other applications.

Have you tested an "env" dump to a(n artifact) file to see if the values are masked or not?

Reply:
No, but that is just one of many ways that a password could accidentally leak out.  It could be difficult to verify all cases, so isolation is the best practice.

With Jenkins traditional jobs, you have to make the environment variables with passwords visible to the build steps.

The Jenkins pipeline provides the method that is designed to only make the sensitive environment variables visible where they are used, so it makes sense to use it and use it that way.